README.md
tw
==

* CUI Twitter Client.
* http://shokai.github.io/tw
* https://rubygems.org/gems/tw


Install
-------

    % gem install tw

for Ruby 2.0 ~ 2.3 users

    % gem install tw --version=1.0.12


Synopsis
--------

show help

    % tw --help

Tweet

    % tw hello hello
    % echo "hello" | tw --pipe
    % tw 'yummy!!' --file=food.jpg

read Timeline

    % tw @username
    % tw @username @user2 @user3
    % tw @username/listname
    % tw --timeline
    % tw --search=ruby
    % tw --stream
    % tw --stream:filter=ruby,java --silent

DM

    % tw --dm
    % tw --dm:to=shokai "hello!"

Reply/Fav/RT

    % tw @shokai --id
    % tw "@shokai wow!!"  --id=334749349588377601
    % tw --id=334749349588377601
    % tw --fav=334749349588377601
    % tw --rt=334749349588377601

Delete

    % tw --del=334749349588377601

Switch Accounts

    % tw --user:add
    open http://twitter.com/oauth/authorize?oauth_token=a1b2cDEF3456gh78
    input PIN Number: 19283746
    add "@user2"
    % tw --user:list
    * shokai
      user2
      user3
    (3 users)
    % tw hello --user=user2
    % tw --user:default=user2
    set default user "@user2"

Format

    % tw --format=json
    % tw --stream --format="@#{user} #{text} - #{url}"


Make Twitter Bot
----------------
stream reply bot

    % tw --silent --stream:filter=BOT_NAME --user=BOT_NAME --format="@#{user} OK" | tw --pipe --user=BOT_NAME


Test
----

    % gem install bundler
    % bundle install
    % bundle exec rake test


Contributing
------------
1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

History.txt
=== 1.3.1 2018-04-15

* use https (#79)
  * thank you for contributing @sachin21

=== 1.3.0 2018-02-18

* 280 chars tweet (#75)

=== 1.2.1 2017-12-10

* Fix deprecated warnings (#77)
* update gems, use twitter v6.2.0, remove yajl-ruby from dependencies (#76)

=== 1.2.0 2017-08-20

* fix --stream option (#73)

=== 1.1.0 2017-07-02

* update rubygems - parallel, rainbow, launchy #72
* fix gemspec for ruby 2.4.x #71

=== 1.0.12 2016-11-30

* use https on tweet URL

=== 1.0.11 2015-10-19

* bugfix search  #65

=== 1.0.10 2015-10-18

* support a timezone  #66
  * thank you for contributing @kaihar4

=== 1.0.9 2014-05-14

* merge multiple lines from STDIN into one tweet

=== 1.0.8 2014-01-28

* fix for rainbow gem v2.0.x
* fix for twitter gem v5.6.x

=== 1.0.7 2014-01-25

* Add --delete option
  * thank you for contributing @janusadm

=== 1.0.6 2014-01-17

* rescue Launchy::CommandNotFoundError  #56

=== 1.0.5 2014-01-16

* use HTTPS for OAuth::Consumer  #55
  * thank you for contributing @ymrl

=== 1.0.4 2014-01-08

* update gem dependencies

=== 1.0.3 2014-01-08

* bugfix for rainbow 1.99.x  #54
* use launchy gem to open OAuth-URL  #53

=== 1.0.2 2013-11-27

* fix syntax error on ruby1.8.7

=== 1.0.1 2013-11-23

* bugfix -user:add  #51

=== 1.0.0 2013-11-21

* fix for twitter gem v5.0.0  #50

=== 0.5.2 2013-09-19

* do not show dialogue with "--yes" option

=== 0.5.1 2013-09-18

* tweet with file  % tw 'yummy!!' --file=food.jpg

=== 0.5.0 2013-08-26

* show RT/Fav count on  % tw --id=1234552234

=== 0.4.9 2013-08-22

* show tweet by status_id  % tw --id=1234543234

=== 0.4.8 2013-08-10

* show Fav/RT count

=== 0.4.7 2013-08-09

* modify DM send dialog
* declare license in gemspec

=== 0.4.6 2013-07-28

* unescape HTML in Tw::Client#show_status

=== 0.4.5 2013-07-27

* fix listname regex, could contain '-'
  * thank you for contributing @takano32

=== 0.4.4 2013-07-14

* silent if --format option

=== 0.4.3 2013-07-02

* notify switch user

=== 0.4.2 2013-05-18

* enable -id=(id) option in pipe-tweet

=== 0.4.1 2013-05-18

* replty to status  % tw "@shokai hello!" --id=1123454321

=== 0.4.0 2013-05-16

* status_id  % tw --id
* RT tweet   % tw --rt=12345136
* Fav tweet  % tw --fav=12345136

=== 0.3.9 2013-03-05

* use http://api.twitter.com to auth
  * thank you for contributing @tdksk

=== 0.3.8 2013-01-04

* use bundler gem template

=== 0.3.7 2012-10-19

* use specify config file : --conf=FILEPATH

=== 0.3.6 2012-10-18

* remove twitter-text gem

=== 0.3.5 2012-10-18

* bugfix

=== 0.3.4 2012-10-18

* add Tw::Client::Stream sample

=== 0.3.3 2012-10-18

* count 140 chars with t.co

=== 0.3.2 2012-10-11

* set config file permission 600

=== 0.3.1 2012-10-11

* pipe DM

=== 0.3.0 2012-10-10

* silent mode option : --silent
* pipe tweet each line from STDIN
* add Tw API sample
* update README

=== 0.2.6 2012-10-10

* bugfix merging users and lists

=== 0.2.5 2012-10-10

* JSON print format : --format=json
* custom print format : --format="#{user} #{text} #{id} #{time}"

=== 0.2.4 2012-10-06

* require json

=== 0.2.3 2012-10-06

* parallel request twitter API

=== 0.2.2 2012-10-06

* unescape HTML

=== 0.2.1 2012-10-05

* auto re-connect stream

=== 0.2.0 2012-10-04

* user stream :  --stream
* filter stream :  --stream:filter=WORD1,WORD2

=== 0.1.1 2012-10-04

* command option detail on help

=== 0.1.0 2012-10-04

* show direct messages : --dm
* create direct message : --dm:to=USERNAME

=== 0.0.3 2012-10-03

* show home timeline : --timeline

=== 0.0.2 2012-10-03

* support multi user accounts : --user=NAME
* tweet from pipe : --pipe
* option : --search=WORD
* show version : --version

=== 0.0.1 2012-09-15

* first release


lib/tw/app/main.rb
require File.expand_path 'opt_parser', File.dirname(__FILE__)
require File.expand_path 'cmds', File.dirname(__FILE__)
require File.expand_path 'render', File.dirname(__FILE__)
require File.expand_path 'helper', File.dirname(__FILE__)

module Tw::App

  def self.new
    Main.new
  end

  class Main
    def initialize
      on_exit do
        exit 0
      end

      on_error do
        exit 1
      end
    end

    def client
      @client ||= Tw::Client.new
    end

    def on_exit(&block)
      if block_given?
        @on_exit = block
      else
        @on_exit.call if @on_exit
      end
    end

    def on_error(&block)
      if block_given?
        @on_error = block
      else
        @on_error.call if @on_error
      end
    end

    def run(argv)
      @args = ArgsParser.parse argv, :style => :equal do
        arg :user, 'user account', :alias => :u
        arg 'user:add', 'add user'
        arg 'user:list', 'show user list'
        arg 'user:default', 'set default user'
        arg :timeline, 'show timeline', :alias => :tl
        arg :dm, 'show direct messages'
        arg 'dm:to', 'create direct message'
        arg :favorite, 'favorite tweet', :alias => :fav
        arg :retweet, 'retweet', :alias => :rt
        arg :delete, 'delete tweet', :alias => :del
        arg :search, 'search public timeline', :alias => :s
        arg :stream, 'show user stream', :alias => :st
        arg :status_id, 'show status_id', :alias => :id
        arg :file, 'upload file'
        arg :pipe, 'pipe tweet'
        arg :format, 'output format', :default => 'text'
        arg :silent, 'silent mode'
        arg :yes, 'do not show dialogue'
        arg :conf, 'config file', :default => Tw::Conf.conf_file
        arg :version, 'show version', :alias => :v
        arg :help, 'show help', :alias => :h

        validate :user, 'invalid user name' do |v|
          v =~ /^[a-zA-Z0-9_]+$/
        end

        validate 'user:default', 'invalid user name' do |v|
          v =~ /^[a-zA-Z0-9_]+$/
        end

        validate 'dm:to', 'invalid user name' do |v|
          v =~ /^[a-zA-Z0-9_]+$/
        end

        validate :file, "file does not exists" do |v|
          File.exists? v
        end
      end

      if @args.has_option? :help
        STDERR.puts "Tw - Twitter client on Ruby v#{Tw::VERSION}"
        STDERR.puts "     http://shokai.github.io/tw"
        STDERR.puts
        STDERR.puts @args.help
        STDERR.puts
        STDERR.puts "e.g."
        STDERR.puts "tweet  tw hello world"
        STDERR.puts "       echo 'hello' | tw --pipe"
        STDERR.puts "       tw 'yummy!!' --file=food.jpg --yes"
        STDERR.puts "read   tw @username"
        STDERR.puts "       tw @username @user2 @user2/listname"
        STDERR.puts "       tw --search=ruby"
        STDERR.puts "       tw --stream"
        STDERR.puts "       tw --stream:filter=ruby,java"
        STDERR.puts "       tw --dm:to=username \"hello!\""
        STDERR.puts "id     tw @shokai --id"
        STDERR.puts "       tw --id=334749349588377601"
        STDERR.puts 'reply  tw "@shokai wow!!" --id=334749349588377601'
        STDERR.puts "       tw --fav=334749349588377601"
        STDERR.puts "       tw --rt=334749349588377601"
        STDERR.puts "       tw --format=json"
        STDERR.puts '       tw --format="@#{user} #{text} - https://twitter.com/#{user}/#{id}"'
        STDERR.puts "delete tw --del=334749349588377601"
        on_exit
      end

      Render.silent = (@args.has_option? :silent or @args.has_param? :format)
      Render.show_status_id = @args[:status_id]
      Tw::Conf.conf_file = @args[:conf]

      regist_cmds

      cmds.each do |name, cmd|
        next unless @args[name]
        cmd.call @args[name], @args
      end

      auth
      if @args.argv.size < 1
        if @args.has_param? :status_id
          client.show_status @args[:status_id]
        else
          Render.display client.mentions, @args[:format]
        end
      elsif all_requests?(@args.argv)
        Render.display Parallel.map(@args.argv, :in_threads => @args.argv.size){|arg|
          if user = username?(arg)
            res = client.user_timeline user
          elsif (user, list = listname?(arg)) != false
            res = client.list_timeline(user, list)
          end
          res
        }, @args[:format]
      else
        message = @args.argv.join(' ')
        tweet_opts = {}
        if (len = message.char_length_with_t_co) > 280
          STDERR.puts "tweet too long (#{len} chars)"
          on_error
        else
          if @args.has_param? :status_id
            client.show_status @args[:status_id]
            puts "--"
            puts "reply \"#{message}\"? (#{len} chars)"
            tweet_opts[:in_reply_to_status_id] = @args[:status_id]
          else
            puts "tweet \"#{message}\"?  (#{len} chars)"
            if @args.has_param? :file
              puts "upload \"#{@args[:file]}\"? (#{File.size @args[:file]} bytes)"
            end
          end
          unless @args.has_option? :yes
            puts '[Y/n]'
            on_exit if STDIN.gets.strip =~ /^n/i
          end
        end
        begin
          if @args.has_param? :file
            client.tweet_with_file message, File.open(@args[:file]), tweet_opts
          else
            client.tweet message, tweet_opts
          end
        rescue => e
          STDERR.puts e.message
        end
      end
    end

    private
    def auth
      return unless @args
      client.auth @args.has_param?(:user) ? @args[:user] : nil
    end

  end
end
